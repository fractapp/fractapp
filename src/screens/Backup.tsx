import React, {useContext, useEffect} from 'react';
import {
  StyleSheet,
  View,
  ScrollView,
  Text,
  FlatList,
  Image,
  TouchableOpacity,
} from 'react-native';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import MaterialIcons from 'react-native-vector-icons/MaterialIcons';
import Dialog from 'storage/Dialog';
import BackupUtils from 'utils/backup';
import db from 'utils/db';
import backupUtil from 'utils/backup';

export const Backup = ({navigation, route}: {navigation: any; route: any}) => {
  const dialogContext = useContext(Dialog.Context);

  const onSuccess = () => {
    navigation.reset({
      index: 0,
      routes: [{name: 'Home'}],
    });
    dialogContext.dispatch(Dialog.close());
  };

  useEffect(() => {
    if (route.params?.isSuccess == null) {
      return;
    }

    const dir =
      route.params.type == BackupUtils.BackupType.File
        ? 'Downloads'
        : backupUtil.GoogleDriveFolder;

    dialogContext.dispatch(
      Dialog.open(
        'Success save wallet',
        `If you lose access to file then you will not be able to restore access to the wallet. File "${route.params?.fileName}.json" saved in "${dir}" directory`,
        onSuccess,
      ),
    );
  }, [route.params?.isSuccess]);

  const backupFile = async () => {
    const seed = await db.getSeed();
    await BackupUtils.backupFile(
      () =>
        navigation.navigate('WalletFileBackup', {
          seed: seed?.split(' '),
          type: BackupUtils.BackupType.File,
          callerScreen: 'Backup',
        }),
      () =>
        dialogContext.dispatch(
          Dialog.open(
            'Open settings',
            'If you want to save a file then open the application settings and give it access to the storage.',
            () => dialogContext.dispatch(Dialog.close()),
          ),
        ),
    );
  };
  const backupGoogleDrive = async () => {
    const seed = await db.getSeed();
    await BackupUtils.backupGoogleDrive(() =>
      navigation.navigate('WalletFileBackup', {
        seed: seed?.split(' '),
        type: BackupUtils.BackupType.GoogleDrive,
        callerScreen: 'Backup',
      }),
    );
  };
  const backupSeed = async () => {
    const seed = await db.getSeed();
    navigation.navigate('SaveSeed', {
      seed: seed?.split(' '),
      onSuccess: () =>
        navigation.reset({
          index: 0,
          routes: [{name: 'Home'}],
        }),
    });
  };

  const menuItems = [
    {
      img: (
        <Image
          source={require('assets/img/google-drive.png')}
          style={{width: 32, height: 32}}
        />
      ),
      title: 'Google disk',
      onClick: () => backupGoogleDrive(),
    },
    {
      img: <MaterialIcons name="folder-open" size={32} color="#888888" />,
      title: 'File to phone',
      onClick: () => backupFile(),
    },
    {
      img: (
        <MaterialCommunityIcons name="content-copy" size={32} color="#888888" />
      ),
      title: 'Backup seed',
      onClick: () => backupSeed(),
    },
  ];
  const renderItem = ({item}: {item: any}) => (
    <TouchableOpacity style={styles.menuItem} onPress={item.onClick}>
      {item.img}
      <Text style={styles.menuTitle}>{item.title}</Text>
    </TouchableOpacity>
  );

  return (
    <View style={styles.profile}>
      <ScrollView showsVerticalScrollIndicator={false} style={{width: '100%'}}>
        <View style={{alignItems: 'center', justifyContent: 'center'}}>
          <Text style={styles.description}>
            Save your wallet in a safe place. If you lose your wallet, you
            cannot restore access to it.
          </Text>
          <FlatList
            style={styles.menu}
            ItemSeparatorComponent={() => <View style={styles.dividingLine} />}
            data={menuItems}
            renderItem={renderItem}
            keyExtractor={(item) => item.title}
          />
        </View>
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  profile: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  dividingLine: {
    alignSelf: 'center',
    width: '100%',
    borderBottomWidth: 1,
    borderBottomColor: '#DADADA',
  },
  description: {
    width: '90%',
    marginTop: 10,
    fontSize: 15,
    fontFamily: 'Roboto-Regular',
    color: '#888888',
  },
  menu: {
    flex: 1,
    width: '88%',
    marginTop: 20,
  },
  menuItem: {
    flexDirection: 'row',
    paddingTop: 10,
    paddingBottom: 10,
  },
  menuTitle: {
    marginLeft: 10,
    fontSize: 17,
    fontFamily: 'Roboto-Regular',
    fontStyle: 'normal',
    color: 'black',
    alignSelf: 'center',
  },
});
