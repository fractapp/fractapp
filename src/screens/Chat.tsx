import React, { useEffect, useState } from 'react';
import { FlatList, Image, Linking, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import { WalletLogo } from 'components/WalletLogo';
import ChatsStore from 'storage/Chats';
import GlobalStore from 'storage/Global';
import stringUtils from 'utils/string';
import StringUtils from 'utils/string';
import backend from 'utils/api';
import { Button, Message } from 'types/message';
import { MessageView } from 'components/MessageView';
import { AddressOnly, Profile, User } from 'types/profile';
import { PaymentMsg } from 'components/PaymentMsg';
import { Currency, getNetwork, toCurrency } from 'types/wallet';
import { randomAsHex } from '@polkadot/util-crypto';
import { useDispatch, useSelector } from 'react-redux';
import UsersStore from 'storage/Users';
import DialogStore from 'storage/Dialog';
import AccountsStore from 'storage/Accounts';
import { Adaptors } from 'adaptors/adaptor';
import { AccountType, Network } from 'types/account';
import ServerInfoStore from 'storage/ServerInfo';

/**
 * Chat screen
 * @category Screens
 */
export const Chat = ({navigation, route}: {navigation: any; route: any}) => {
  const globalState: GlobalStore.State = useSelector((state: any) => state.global);
  const usersState: UsersStore.State = useSelector((state: any) => state.users);
  const accountsState: AccountsStore.State = useSelector((state: any) => state.accounts);
  const chatsState: ChatsStore.State = useSelector((state: any) => state.chats);
  const serverInfoState: ServerInfoStore.State = useSelector((state: any) => state.serverInfo);

  const dispatch = useDispatch();

  const chatId: string = route.params.chatId;

  const chatInfo = chatsState.chatsInfo[chatId]!;
  const user: User = usersState.users[chatId]!;

  const [notificationCount] = useState(chatInfo.notificationCount);
  const [messages, setMessages] = useState<Array<Message>>([]);
  const [lengthOffset, setLengthOffset] = useState<number>(0);

  const onPressChatBtn = async (msgId: string, btn: Button) => {
    switch (btn.action) {
      case '/enterAmount':
        if (btn.arguments.length < 3) {
          return;
        }

        const currencyEnterAmount: Currency = toCurrency(btn.arguments[0]);

        navigation.navigate('EnterAmount', {
          chatId: chatId,
          msgId: msgId,
          isChatBotRequest: true,
          isUSDMode: true,
          value: '',
          currency: currencyEnterAmount,
          args: btn.arguments.slice(1, btn.arguments.length),
        });
        break;
      case '/openUrl':
        if (btn.arguments.length < 1) {
          return;
        }
        await Linking.openURL(btn.arguments[0]); //TODO: next release alert (any site)
        break;
      case '/broadcast':
        try {
          if (btn.arguments.length < 2) {
            return;
          }

          const tx = JSON.parse(btn.arguments[0]);
          const currency: Currency = toCurrency(btn.arguments[1]);

          const network: Network = getNetwork(currency);
          const api = Adaptors.get(network)!;

          dispatch(DialogStore.actions.showConfirmTxInfo(
            await api.parseTx((user.value as Profile),
              accountsState.accounts[AccountType.Main][currency],
              msgId,
              btn.arguments.slice(2, btn.arguments.length),
              tx,
              serverInfoState.prices[currency])
            )
          );
        } catch (e) {
          console.log('broadcast err: ' + e);
          return;
        }
        break;
      default:
        let msg = {
          id: 'answer-' + randomAsHex(32),
          value: btn.value,
          action: btn.action,
          args: btn.arguments,
          rows: [],
          timestamp: Date.now(),
          sender: globalState.profile!.id,
          receiver: chatInfo.id,
          hideBtn: true,
        };
        const timestamp = await backend.sendMsg({
          version: 1,
          value: btn.value,
          action: btn.action,
          receiver: chatInfo.id,
          args: btn.arguments,
        });
        if (timestamp == null) {
          return;
        }

        msg.timestamp = timestamp;

        dispatch(ChatsStore.actions.hideBtns({
          chatId: chatInfo.id,
          msgId: msgId,
        }));
        dispatch(ChatsStore.actions.addMessages([{
          chatId: chatInfo.id,
          msg: msg,
        }]));
        break;
    }
  };

  useEffect(() => {
    navigation.setOptions({
      title: stringUtils.formatNameOrAddress(user.title),
      headerRight: () => {
        return (
          <TouchableOpacity
            onPress={() =>
              navigation.navigate('ProfileInfo', {
                userId: chatInfo.id,
              })
            }
          >
            {user.isAddressOnly ?
              <WalletLogo
                currency={(user.value as AddressOnly).currency}
                size={45}
              /> :
              <Image
                source={{
                  uri: backend.getImgUrl((user.value as Profile).id, (user.value as Profile).lastUpdate),
                }}
                width={45}
                height={45}
                style={{ width: 45, height: 45, borderRadius: 25 }}
              />
            }
          </TouchableOpacity>
        );
      },
      headerRightContainerStyle: {
        marginRight: 20,
      },
    });
  }, []);

  useEffect(() => {
    const newMessages =
      Object.keys(chatsState.chats[chatInfo.id].messages)
      .map((key) => chatsState.chats[chatInfo.id].messages[key])
      .sort((a, b) => b.timestamp - a.timestamp);

    setMessages(newMessages);
    setLengthOffset(messages.length === 0 ? 0 : lengthOffset + (messages.length - newMessages.length));
  }, [chatsState.chats[chatInfo.id].messages]);

  useEffect(() => {
    if (chatInfo.notificationCount !== 0) {
      dispatch(ChatsStore.actions.removeNotification(chatInfo.id));
    }
  }, [chatInfo.notificationCount]);

  const renderItem = ({item, index}: {item: Message; index: number}) => {
    let line;
    if (notificationCount !== 0 && index === notificationCount - lengthOffset - 1) {
      line = (
        <View
          style={{
            marginTop: 10,
            marginBottom: 10,
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'center',
          }}>
          <Text style={styles.newMsgText}>
            {StringUtils.texts.UnreadMessagesTitle}
          </Text>
        </View>
      );
    }

    return (
      <View
        key={item.id}
        style={{
        paddingLeft: 15,
        paddingRight: 15,
        scaleY: -1,
      }}>
        {line}
        {
          item.action === '/tx' &&
          item.args.length >= 2 &&
          chatsState.transactions[toCurrency(item.args[0])].transactionById[item.args[1]] ?
            (() => {
              const tx = chatsState.transactions[toCurrency(item.args[0])].transactionById[item.args[1]];
              return <PaymentMsg
                onPress={() =>
                  navigation.navigate('TransactionDetails', {
                    transaction: tx,
                  })}
                tx={{
                  id: tx.id,
                  hash: tx.hash,
                  fullValue: tx.fullValue,
                  userId: tx.userId,
                  address: tx.address,
                  action: tx.action,
                  currency: tx.currency,
                  txType: tx.txType,
                  timestamp: tx.timestamp,
                  value: tx.value,
                  planckValue: tx.planckValue,
                  usdValue: tx.usdValue,
                  fee: tx.fee,
                  planckFee: tx.planckFee,
                  usdFee: tx.usdFee,
                  status: tx.status,
                }} />;
            })() :
            (<MessageView key={item.id} message={item} isOwner={item.sender === globalState.profile!.id}
                          onPressChatBtn={onPressChatBtn} />)
        }
      </View>
    );
  };

  return (
    <View style={styles.chat}>
      {<FlatList
        style={[{scaleY: -1 }, styles.messages]}
        getItemLayout={ (data, index) => (
          { length: 50, offset: 50 * index, index }
        )}
        windowSize={5}
        data={messages}
        renderItem={renderItem}
        keyExtractor={(item) => item.id}
        ListHeaderComponent={<View style={{marginBottom: user.isAddressOnly || !(user.value as Profile).isChatBot ? 80 : 50}}  />}
      />}
      {((user.isAddressOnly || !(user.value as Profile).isChatBot) &&
        <TouchableOpacity
          style={styles.sendBox}
          onPress={() =>
            !user.isAddressOnly
              ? navigation.navigate('SelectWallet', {
                chatId: chatInfo.id,
              })
              : navigation.navigate('Send', {
                isEditable: false,
                chatId: chatInfo.id,
                currency: (user.value as AddressOnly).currency,
              })
          }>
          <Image
            source={require('assets/img/send.png')}
            style={{width: 16, height: 25}}
          />
        </TouchableOpacity>
      )}
      {(!user.isAddressOnly && (user.value as Profile).isChatBot) && messages.length == 0 &&
        <TouchableOpacity
          style={styles.sendBox}
          onPress={() => {
            const msg = {
              id: 'answer-' + randomAsHex(32),
              value: 'Start',
              action: '/init',
              args: [],
              rows: [],
              timestamp: Date.now(),
              sender: globalState.profile!.id,
              receiver: chatInfo.id,
              hideBtn: true,
            };
            backend.sendMsg({
              version: 1,
              value: msg.value,
              action: msg.action,
              receiver: chatInfo.id,
              args: msg.args,
            }).then((timestamp) => {
              if (timestamp != null) {
                msg.timestamp = timestamp;
                dispatch(ChatsStore.actions.addMessages([{
                  chatId: chatInfo.id,
                  msg: msg,
                }]));
              }
            });
          }}
        >
          <Text style={styles.startBtnText}>Start</Text>
        </TouchableOpacity>
      }
    </View>
  );
};

const styles = StyleSheet.create({
  buttonText: {
    fontSize: 17,
    fontFamily: 'Roboto-Regular',
    fontStyle: 'normal',
    fontWeight: 'normal',
    color: 'white',
  },
  sendBox: {
    position: 'absolute',
    right: 10,
    bottom: 10,
    height: 60,
    width: 60,
    borderRadius: 50,
    backgroundColor: '#2AB2E2',
    alignItems: 'center',
    justifyContent: 'center',
  },
  chat: {
    flex: 1,
    alignItems: 'center',
    flexDirection: 'column',
  },
  messages: {
    width: '100%',
  },
  dividingLine: {
    width: '100%',
    height: 0.5,
    backgroundColor: '#C5C5C5',
  },
  newMsgDividingLine: {
    width: '15%',
    height: 1,
    borderColor: '#2AB2E2',
    borderBottomWidth: 1,
  },
  newMsgText: {
    textAlign: 'center',
    width: '100%',
    backgroundColor: '#51caf5',
    fontSize: 15,
    fontFamily: 'Roboto-Regular',
    fontStyle: 'normal',
    fontWeight: 'normal',
    color: 'white',
  },
  startBtnText: {
    fontSize: 15,
    fontFamily: 'Roboto-Regular',
    fontStyle: 'normal',
    fontWeight: 'normal',
    color: 'white',
  },
});
