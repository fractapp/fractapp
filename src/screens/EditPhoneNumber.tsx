import React, {useEffect, useState} from 'react';
import {
  StyleSheet,
  TextInput,
  Text,
  View,
  TouchableOpacity,
} from 'react-native';
import {SuccessButton} from 'components/SuccessButton';
import BackendApi from 'utils/api';
import Dialog from 'storage/Dialog';
import backend from 'utils/api';
import StringUtils from 'utils/string';
import { useDispatch } from 'react-redux';
// @ts-ignore
import {
  formatPhoneNumberIntl,
  isValidPhoneNumber,
  parsePhoneNumber,
  getCountryCallingCode,
} from 'react-phone-number-input';
// @ts-ignore
import en from 'react-phone-number-input/locale/en';

/**
 * Screen with editing phone number in fractapp
 * @category Screens
 */
export const EditPhoneNumber = ({
  navigation,
  route,
}: {
  navigation: any;
  route: any;
}) => {
  const dispatch = useDispatch();

  const [countryCode, setCountryCode] = useState<string>('US');
  const [number, setNumber] = useState<string>('1');
  const [countryCodeLength, setCountryCodeLength] = useState<number>(1);
  const [countryName, setCountryName] = useState<string>('United States');
  const selectedCountryCode = route.params?.selectedCountryCode;

  const onSuccess = async () => {
    if (!isValidPhoneNumber('+' + number)) {
      dispatch(
        Dialog.actions.showDialog({
          title: StringUtils.texts.InvalidPhoneNumberTitle,
          text: StringUtils.texts.InvalidPhoneNumberText,
          onPress: () => dispatch(Dialog.actions.hideDialog()),
        }),
      );
      return;
    }

    try {
      BackendApi.sendCode(
        number,
        BackendApi.CodeType.Phone
      );
    } catch (e) {
      console.log(e);
    }

    navigation.navigate('ConfirmCode', {
      value: number,
      type: BackendApi.CodeType.Phone,
    });
  };

  useEffect(() => {
    if (selectedCountryCode != null) {
      return;
    }

    (async () => {
      const local = await backend.getLocalByIp();
      const numberCountryCode = getCountryCallingCode(local);
      setCountryCodeLength(numberCountryCode.length);
      setCountryCode(local);
      setCountryName(en[local]);

      setNumber(numberCountryCode);
    })();
  }, []);
  useEffect(() => {
    if (selectedCountryCode == null) {
      return;
    }

    const numberCountryCode = getCountryCallingCode(selectedCountryCode);

    setCountryCodeLength(numberCountryCode.length);
    setCountryCode(selectedCountryCode);
    setCountryName(en[selectedCountryCode]);

    setNumber(numberCountryCode);
  }, [selectedCountryCode]);
  useEffect(() => {
    navigation.setOptions({
      headerRight: () => {
        return <SuccessButton size={35} onPress={onSuccess} />;
      },
    });

    if (number.length < countryCodeLength) {
      setCountryCode('');
      setCountryName(StringUtils.texts.edit.invalidPhoneNumber);
    }

    if (countryCode !== '') {
      return;
    }
    if (isValidPhoneNumber('+' + number)) {
      const info = parsePhoneNumber('+' + number);
      if (info?.country == null) {
        return;
      }
      setCountryCode(info.country);
      setCountryName(en[info.country!]);
    } else {
      setCountryName(StringUtils.texts.edit.invalidPhoneNumber);
    }
  }, [number]);

  return (
    <View style={styles.box}>
      <TouchableOpacity
        testID={'selectCountryBtn'}
        style={styles.countryInput}
        onPress={() => navigation.navigate('SelectCountry')}>
        <Text style={styles.title}>{StringUtils.texts.edit.countryTitle}</Text>
        <Text style={[styles.value, {marginTop: 6}]}>{countryName}</Text>
      </TouchableOpacity>
      <View style={styles.phoneInput}>
        <Text style={styles.title}>{StringUtils.texts.edit.phoneTitle}</Text>
        <View
          style={{
            flexDirection: 'row',
            alignContent: 'center',
            marginTop: 0,
          }}>
          <Text style={[styles.value, {alignSelf: 'center'}]}>+</Text>
          <TextInput
            style={[
              styles.value,
              {
                width: '100%',
              },
            ]}
            value={number}
            onChangeText={(text) => {
              if (isValidPhoneNumber('+' + text)) {
                setNumber(formatPhoneNumberIntl('+' + text).substring(1));
              } else {
                setNumber(text);
              }
            }}
            placeholder={''}
            keyboardType={'phone-pad'}
            placeholderTextColor={'#BFBDBD'}
            autoCompleteType={'tel'}
            textContentType={'telephoneNumber'}
            secureTextEntry={false}
          />
        </View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  box: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  countryInput: {
    width: '85%',
    marginTop: 20,
    paddingBottom: 5,
    borderColor: '#CCCCCC',
    borderBottomWidth: 1,
  },
  phoneInput: {
    width: '85%',
    marginTop: 20,
    borderColor: '#CCCCCC',
    borderBottomWidth: 1,
  },
  title: {
    fontSize: 16,
    fontFamily: 'Roboto-Regular',
    fontStyle: 'normal',
    fontWeight: 'normal',
    color: '#888888',
  },
  value: {
    fontSize: 16,
    fontFamily: 'Roboto-Regular',
    fontStyle: 'normal',
    fontWeight: 'normal',
    color: 'black',
  },
});
