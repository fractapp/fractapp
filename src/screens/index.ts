export {ImportWallet} from './ImportWallet';
export {SaveSeed} from './SaveSeed';
export {SaveWallet} from './SaveWallet';
export {SettingWallet} from './SettingWallet';
export {Start} from './Start';
export {WalletDetails} from './WalletDetails';
export {WalletFileBackup} from './WalletFileBackup';
export {Wallets} from './Wallets';
export {Settings} from './Settings';
export {ConfirmSaveSeed} from './ConfirmSaveSeed';
export {ImportSeed} from './ImportSeed';
export {WalletFileImport} from './WalletFileImport';
export {GoogleDrivePicker} from './GoogleDrivePicker';
export {Receive} from './Receive';
export {Milestone} from './Milestone';
export {TransactionDetails} from './TransactionDetails';
export {SplashScreen} from './SplashScreen';
export {Backup} from './Backup';
export {NewPassCode} from './NewPassCode';
export {VerifyPassCode} from './VerifyPassCode';
export {EditProfile} from './EditProfile';
export {EditPhoneNumber} from './EditPhoneNumber';
export {SelectCountry} from './SelectCountry';
export {ConfirmCode} from './ConfirmCode';
export {EditEmail} from './EditEmail';
export {Navigation} from './Navigation';
export {Chats} from './Chats';
export {Chat} from './Chat';
export {Connecting} from './Connecting';
export {EditUsername} from './EditUsername';
export {EditName} from './EditName';
export {Search} from './Search';
export {SelectWallet} from './SelectWallet';
export {Send} from './Send';
export {EnterAmount} from './EnterAmount';
export {SendWithEnterAddress} from './SendWithEnterAddress';
