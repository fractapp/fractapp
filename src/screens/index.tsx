export { default as ImportWallet } from './ImportWallet';
export { default as SaveSeed } from './SaveSeed';

export { default as SaveWallet } from './SaveWallet';
export { default as SettingWallet } from './SettingWallet';
export { default as Start } from './Start';

export { default as WalletDetails } from './WalletDetails';

export { default as WalletEncryption } from './WalletEncryption';
export { default as Wallets } from './Wallets';
export { ConfirmSaveSeed } from './ConfirmSaveSeed'