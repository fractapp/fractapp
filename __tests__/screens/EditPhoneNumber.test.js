import React, {useState} from 'react';
import renderer from 'react-test-renderer';
import {EditPhoneNumber} from 'screens/EditPhoneNumber';
import BackendApi from 'utils/backend';
import {fireEvent, render} from '@testing-library/react-native';
import StringUtils from 'utils/string';
jest.mock('storage/DB', () => ({}));
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
  Linking: jest.fn(() => ({
    openURL: jest.fn(),
  })),
}));
jest.mock('utils/backend', () => ({
  getLocalByIp: jest.fn(),
}));
jest.mock('react-native-i18n', () => ({
  t: (value) => value,
}));

it('Test view with empty number', async () => {
  useState.mockImplementation((init) => [init, jest.fn()]);

  BackendApi.getLocalByIp.mockReturnValueOnce('AD');
  const tree = renderer
    .create(
      <EditPhoneNumber
        navigation={{setOptions: jest.fn()}}
        route={{
          params: {},
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Test useEffect with empty number', async () => {
  const setCountryCode = jest.fn();
  const setNumber = jest.fn();
  const setCountryCodeLength = jest.fn();
  const setCountryName = jest.fn();

  useState.mockImplementationOnce((init) => [init, setCountryCode]);
  useState.mockImplementationOnce((init) => [init, setNumber]);
  useState.mockImplementationOnce((init) => [init, setCountryCodeLength]);
  useState.mockImplementationOnce((init) => [init, setCountryName]);

  BackendApi.getLocalByIp.mockReturnValueOnce('AD');
  const component = await render(
    <EditPhoneNumber
      navigation={{setOptions: jest.fn()}}
      route={{
        params: {},
      }}
    />,
  );

  expect(setCountryCodeLength).toBeCalledWith(3);
  expect(setCountryCode).toBeCalledWith('AD');
  expect(setCountryName).toBeCalledWith('Andorra');
  expect(setNumber).toBeCalledWith('376');
});

it('Test useEffect with selectedCountryCode', async () => {
  const setCountryCode = jest.fn();
  const setNumber = jest.fn();
  const setCountryCodeLength = jest.fn();
  const setCountryName = jest.fn();

  useState.mockImplementationOnce((init) => [init, setCountryCode]);
  useState.mockImplementationOnce((init) => [init, setNumber]);
  useState.mockImplementationOnce((init) => [init, setCountryCodeLength]);
  useState.mockImplementationOnce((init) => [init, setCountryName]);

  BackendApi.getLocalByIp.mockReturnValueOnce('AG');
  const component = await render(
    <EditPhoneNumber
      navigation={{setOptions: jest.fn()}}
      route={{
        params: {selectedCountryCode: 'AR'},
      }}
    />,
  );

  expect(setCountryCodeLength).toBeCalledWith(2);
  expect(setCountryCode).toBeCalledWith('AR');
  expect(setCountryName).toBeCalledWith('Argentina');
  expect(setNumber).toBeCalledWith('54');
});

it('Test useEffect with number', async () => {
  const setCountryCode = jest.fn();
  const setNumber = jest.fn();
  const setCountryCodeLength = jest.fn();
  const setCountryName = jest.fn();

  useState.mockImplementationOnce((init) => ['', setCountryCode]);
  useState.mockImplementationOnce((init) => ['13192000228', setNumber]);
  useState.mockImplementationOnce((init) => [init, setCountryCodeLength]);
  useState.mockImplementationOnce((init) => [init, setCountryName]);

  BackendApi.getLocalByIp.mockReturnValueOnce('');
  const component = await render(
    <EditPhoneNumber
      navigation={{setOptions: jest.fn()}}
      route={{
        params: {},
      }}
    />,
  );

  expect(setCountryCode).toBeCalledWith('US');
  expect(setCountryName).toBeCalledWith('United States');
});

it('Test useEffect with invalid number', async () => {
  const setCountryCode = jest.fn();
  const setNumber = jest.fn();
  const setCountryCodeLength = jest.fn();
  const setCountryName = jest.fn();

  useState.mockImplementationOnce((init) => ['', setCountryCode]);
  useState.mockImplementationOnce((init) => ['1319200022', setNumber]);
  useState.mockImplementationOnce((init) => [init, setCountryCodeLength]);
  useState.mockImplementationOnce((init) => [init, setCountryName]);

  BackendApi.getLocalByIp.mockReturnValueOnce('');
  const component = await render(
    <EditPhoneNumber
      navigation={{setOptions: jest.fn()}}
      route={{
        params: {},
      }}
    />,
  );

  expect(setCountryName).toBeCalledWith(
    StringUtils.texts.edit.invalidPhoneNumber,
  );
});

it('Test click selectCountry', async () => {
  const navigate = jest.fn();
  const setCountryCode = jest.fn();
  const setNumber = jest.fn();
  const setCountryCodeLength = jest.fn();
  const setCountryName = jest.fn();

  useState.mockImplementationOnce((init) => ['', setCountryCode]);
  useState.mockImplementationOnce((init) => ['1319200022', setNumber]);
  useState.mockImplementationOnce((init) => [init, setCountryCodeLength]);
  useState.mockImplementationOnce((init) => [init, setCountryName]);

  BackendApi.getLocalByIp.mockReturnValueOnce('');
  const component = await render(
    <EditPhoneNumber
      navigation={{navigate: navigate, setOptions: jest.fn()}}
      route={{
        params: {},
      }}
    />,
  );

  fireEvent.press(component.getByTestId('selectCountryBtn'));
  expect(navigate).toBeCalledWith('SelectCountry');
});

/*it('Test onSuccess', async () => {
  const navigate = jest.fn();
  const setNumber = jest.fn();
  useState.mockImplementationOnce((init) => ['1319200022', setNumber]);

  BackendApi.getLocalByIp.mockReturnValueOnce('');
  const component = await render(
    <EditPhoneNumber
      navigation={{navigate: navigate, setOptions: jest.fn()}}
      route={{
        params: {},
      }}
    />,
  );
  fireEvent.press(component.getByTestId('successBtn'));
  expect(navigate).toBeCalledWith('ConfirmCode');
});
*/