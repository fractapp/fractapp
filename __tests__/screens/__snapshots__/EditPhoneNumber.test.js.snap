// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test view with empty number 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#CCCCCC",
        "marginTop": 20,
        "opacity": 1,
        "paddingBottom": 5,
        "width": "85%",
      }
    }
    testID="selectCountryBtn"
  >
    <Text
      style={
        Object {
          "color": "#888888",
          "fontFamily": "Roboto-Regular",
          "fontSize": 16,
          "fontStyle": "normal",
          "fontWeight": "normal",
        }
      }
    >
      [missing "en.edit.country_title" translation]
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "black",
            "fontFamily": "Roboto-Regular",
            "fontSize": 16,
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {
            "marginTop": 6,
          },
        ]
      }
    >
      United States
    </Text>
  </View>
  <View
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#CCCCCC",
        "marginTop": 20,
        "width": "85%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#888888",
          "fontFamily": "Roboto-Regular",
          "fontSize": 16,
          "fontStyle": "normal",
          "fontWeight": "normal",
        }
      }
    >
      [missing "en.edit.phone_title" translation]
    </Text>
    <View
      style={
        Object {
          "alignContent": "center",
          "flexDirection": "row",
          "marginTop": 0,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "black",
              "fontFamily": "Roboto-Regular",
              "fontSize": 16,
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {
              "alignSelf": "center",
            },
          ]
        }
      >
        +
      </Text>
      <TextInput
        allowFontScaling={true}
        autoCompleteType="tel"
        keyboardType="phone-pad"
        onChangeText={[Function]}
        placeholder=""
        placeholderTextColor="#BFBDBD"
        rejectResponderTermination={true}
        secureTextEntry={false}
        style={
          Array [
            Object {
              "color": "black",
              "fontFamily": "Roboto-Regular",
              "fontSize": 16,
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {
              "width": "100%",
            },
          ]
        }
        textContentType="telephoneNumber"
        underlineColorAndroid="transparent"
        value="1"
      />
    </View>
  </View>
</View>
`;
